crates/symbolicator-service/src/cache.rs

```diff
@@ -119,6 +120,174 @@ impl CacheStatus {
+        F: FnOnce(T) -> Result<U, E>,
+    {
+        match self {
+            Self::AllGood(entry) => match f(entry) {
+                Ok(new_entry) => CacheEntry::AllGood(new_entry),
+                Err(e) => {
+                    tracing::error!(error = %e);
+                    CacheEntry::InternalError
+                }
+            },
+            Self::NotFound => CacheEntry::NotFound,
+            Self::PermissionDenied(details) => CacheEntry::PermissionDenied(details),
+            Self::Timeout(duration) => CacheEntry::Timeout(duration),
+            Self::DownloadError(details) => CacheEntry::DownloadError(details),
+            Self::Malformed(details) => CacheEntry::Malformed(details),
+            Self::InternalError => CacheEntry::InternalError,
+        }
+    }
+
+    pub fn all_good(self) -> Option<T> {
+        match self {
+            Self::AllGood(entry) => Some(entry),
+            _ => None,
+        }
+    }
+
+    pub fn is_not_found(&self) -> bool {
+        matches!(self, Self::NotFound)
+    }
+}
+
+impl CacheEntry<ByteView<'static>> {
+    pub fn from_bytes(bytes: ByteView<'static>) -> Self {
+        if let Some(raw_message) = bytes.strip_prefix(Self::PERMISSION_DENIED_MARKER) {
+            let err_msg = String::from_utf8_lossy(raw_message);
+            Self::PermissionDenied(err_msg.into_owned())
+        } else if let Some(raw_duration) = bytes.strip_prefix(Self::TIMEOUT_MARKER) {
+            let raw_duration = String::from_utf8_lossy(raw_duration);
+            match parse_duration(&raw_duration) {
+                Ok(duration) => Self::Timeout(duration),
+                Err(e) => {
+                    tracing::error!(error = %e, "Failed to read timeout duration");
+                    Self::InternalError
+                }
+            }
+        } else if let Some(raw_message) = bytes.strip_prefix(Self::DOWNLOAD_ERROR_MARKER) {
+            let err_msg = String::from_utf8_lossy(raw_message);
+            Self::DownloadError(err_msg.into_owned())
+        } else if let Some(raw_message) = bytes.strip_prefix(Self::MALFORMED_MARKER) {
+            let err_msg = String::from_utf8_lossy(raw_message);
+            Self::Malformed(err_msg.into_owned())
+        } else if bytes.as_ref() == Self::INTERNAL_ERROR_MARKER {
+            Self::InternalError
+        } else if bytes.is_empty() {
+            Self::NotFound
+        } else {
+            Self::AllGood(bytes)
+        }
+    }
+}
+
+impl From<(CacheStatus, ByteView<'static>)> for CacheEntry<ByteView<'static>> {
+    fn from((status, content): (CacheStatus, ByteView<'static>)) -> Self {
+        match status {
+            CacheStatus::Positive => Self::AllGood(content),
+            CacheStatus::Negative => Self::NotFound,
+            CacheStatus::Malformed(s) => Self::Malformed(s),
+            CacheStatus::CacheSpecificError(message) => {
+                if let Some(details) = message.strip_prefix("missing permissions for file") {
```

Comment:
can we use whatever constant is used inside the `DownloadError`, or is it not that simple?

Reply:
`DownloadError` uses `&Self::Permissions.to_string()`. We could use that, but tying another completely unrelated type into this conversion seems wrong to me.

Reply:
https://github.com/getsentry/symbolicator/blob/7bad96c1908a9772d3d259a9808ab26f1417bf98/crates/symbolicator-service/src/services/download/mod.rs#L104-L105

lol, current code uses `to_string()` just to check for that condition